Sources/SwiftCliCore/OffsetUtility.swift

```diff
@@ -0,0 +1,87 @@
+import Foundation
+
+public struct RowInfo: Equatable {
+    var row: [String]
+    var rowNumber: Int
+    var bottomLine: String = ""
+    var stitchSymbols: String = ""
+    var width: Int = 0
+    var leftIncDec: Int = 0
+    var rightIncDec: Int = 0
+    // This is about the number of "empty stitches" that should be added on the left to account for future increases
+    var leftOffset: Int = 0
+    lazy var leftOffsetString: String = String(repeating: "  ", count: leftOffset)
+    lazy var offsetBottomLine: String = leftOffsetString + bottomLine
+    lazy var offsetStitchSymbols: String = leftOffsetString + stitchSymbols
+    lazy var totalRow: String = offsetStitchSymbols + offsetBottomLine
+}
+
+public class OffsetUtility {
+    public init() {}
+
+    public func gatherAllMetaData(stitchArray: [[String]]) -> [RowInfo] {
+        var allRowsMetaData = [] as [RowInfo]
+        let rowsNum = stitchArray.count
+        for row in 0..<(rowsNum) {
+            let newRowMetadata = makeRowMetadata(stitchRow: stitchArray[row], rowNumber: row)
+            allRowsMetaData.append(newRowMetadata)
+            if newRowMetadata.leftIncDec != 0 {
+                for prevRow in (0...(row-1)).reversed() {
+                    allRowsMetaData[prevRow].leftOffset += newRowMetadata.leftIncDec
+                }
+            }
+        }
+        return(allRowsMetaData)
+    }
+
+
+    public func makeRowMetadata(stitchRow: [String], rowNumber: Int) -> RowInfo {
+        var rowData = RowInfo(row: stitchRow, rowNumber: rowNumber)
+        rowData.leftIncDec = findLeftChanges(stitchRow: stitchRow)
+        rowData.rightIncDec = findRightChanges(stitchRow: stitchRow)
+        rowData.width = stitchRow.count
+        if rowNumber == 0 {
+            rowData.bottomLine = ChartConstructor().makeBottomRow(width: rowData.width)
+        } else {
+            rowData.bottomLine = ChartConstructor().makeMiddleRow(width: rowData.width,
+                                                                  rDiff: rowData.rightIncDec,
+                                                                  lDiff: rowData.leftIncDec)
+        }
+        rowData.stitchSymbols = ChartConstructor().makeStitchRow(row: rowData.row)
+        return(rowData)
+    }
+
+    public func findCenter(stitchRow: [String]) -> Int {
+        let center = floor(Double(stitchRow.count/2))
+    return(Int(center))
+    }
+
+    public func findLeftChanges(stitchRow: [String]) -> Int {
+        let center = findCenter(stitchRow: stitchRow)
+        var totalLeftChange = 0
+        for stitch in 0..<(center) {
+            if increases.contains(stitchRow[stitch]) {
+                totalLeftChange += 1
+            }
+            if decreases.contains(stitchRow[stitch]) {
+                totalLeftChange += -1
+            }
+        }
```

Comment:
Rather than looping, you could take advantage of the `filter` method: https://developer.apple.com/documentation/swift/sequence/3018365-filter

```swift
let leftHalf = 0..<(center)
leftIncreases = leftHalf.filter { increases.contains($0) }
leftDecreases = leftHalf.filter { decreases.contains($0) }
return leftIncreases.count - leftDecreases.count
```

Reply:
Also it took me a while to figure out where `increases` and `decreases` came from. You might want to take the suggestion from the style guide and namespace these constants by making them cases of an enum as indicated here: https://github.com/raywenderlich/swift-style-guide#constants

Note that the style guide also recommends using `static let` when setting the value.

Reply:
The enum thing makes much more sense, that way I could also add a Raw Value to represent the number of increases or decreases the stitch represents, as some represent multiple decreases, like k3tog.

However, I'm torn between doing this and creating more complex types for the stitches - perhaps a stitch class that includes the name of the stitch, the increase or decrease value, and the string of how it is printed. This could be useful for later adding values like "k5", for example - there could be instructions for how to parse user input that way. 

I think I might go ahead and create the Stitch class type because I see it as being easier to expand upon later, rather than continuing to have a set of allowed stitches, and then a dictionary with the symbols associated with each stitch, and a list of which stitches represent increases and decreases.

Reply:
That's a good idea! If you do create a `Stitch` class, you could give it a property like `stitchSymbol`, which would eliminate the need for a dictionary. You could then just have a set of allowed stitches the members of which are instances of the `Stitch` class.

Reply:
Actually, upon reading more, I think I can use the enum class too for this. Sorry for the contradictory messages!

Reply:
No worries! Looking forward to what you come up with.